src/kernels/MIOpenSparseSoftmaxCrossEntropyWithLogits.cpp

```diff
@@ -0,0 +1,280 @@
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#ifndef T_TYPE
+#define T_TYPE int
+#endif
+
+#ifndef DTYPE_LOWEST
+#define DTYPE_LOWEST 0
+#endif
+
+#include "float_types.h"
+#include "tensor_view.hpp"
+
+template <typename T, typename Ta>
+__device__ void sparseSoftmaxCrossEntropyWithLogitsForward(const T* input,
+                                                           const Ta* target,
+                                                           T* output,
+                                                           T* backprop,
+                                                           uint64_t num_class,
+                                                           tensor_view_t<2> input_tv,
+                                                           tensor_view_t<1> target_tv,
+                                                           tensor_view_t<1> output_tv,
+                                                           tensor_view_t<2> backprop_tv)
+{
+    uint64_t gid = blockIdx.x;
+    uint64_t lid = threadIdx.x;
+
+    __shared__ FLOAT_ACCUM lmax[LOCAL_SIZE], lsum[LOCAL_SIZE];
+    lmax[lid] = DTYPE_LOWEST;
+    lsum[lid] = 0.0f;
+    __shared__ uint64_t label;
+
+    if(lid == 0)
+    {
+        label = static_cast<uint64_t>(target[target_tv.get_tensor_view_idx({gid})]);
+    }
+
+    for(uint64_t i = lid; i < num_class; i += LOCAL_SIZE)
+    {
+        FLOAT_ACCUM val = CVT_FLOAT2ACCUM(input[input_tv.get_tensor_view_idx({gid, i})]);
+        lmax[lid]       = max(lmax[lid], val);
+    }
+    __syncthreads();
+
+    for(uint64_t i = LOCAL_SIZE >> 1; i > 0; i >>= 1)
+    {
+        if(lid < i)
+        {
+            lmax[lid] = max(lmax[lid], lmax[lid + i]);
+        }
+        __syncthreads();
+    }
+
+    for(uint64_t i = lid; i < num_class; i += LOCAL_SIZE)
+    {
+        FLOAT_ACCUM val = CVT_FLOAT2ACCUM(input[input_tv.get_tensor_view_idx({gid, i})]);
+        lsum[lid] += exp(val - lmax[0]);
+    }
+    __syncthreads();
+
+    for(uint64_t i = LOCAL_SIZE >> 1; i > 0; i >>= 1)
+    {
+        if(lid < i)
+        {
+            lsum[lid] += lsum[lid + i];
+        }
+        __syncthreads();
+    }
+
+    for(uint64_t i = lid; i < num_class; i += LOCAL_SIZE)
+    {
+        if(i == label)
+        {
+            FLOAT_ACCUM val = CVT_FLOAT2ACCUM(input[input_tv.get_tensor_view_idx({gid, i})]);
+            output[output_tv.get_tensor_view_idx({gid})] =
+                CVT_ACCUM2FLOAT(log(lsum[0]) - val + lmax[0]);
+        }
+    }
+
+    for(uint64_t i = lid; i < num_class; i += LOCAL_SIZE)
+    {
+        FLOAT_ACCUM val = CVT_FLOAT2ACCUM(input[input_tv.get_tensor_view_idx({gid, i})]);
+        FLOAT_ACCUM backprop_val =
+            (i == label) ? exp(val - lmax[0]) / lsum[0] - 1.0f : exp(val - lmax[0]) / lsum[0];
+
+        backprop[backprop_tv.get_tensor_view_idx({gid, i})] = CVT_ACCUM2FLOAT(backprop_val);
+    }
```

Comment:
Should those 2 for loop merged? Maybe we can cut off 1 call for getting `input` value. I know that ths is moDNN implementation, but it's quite obvious.